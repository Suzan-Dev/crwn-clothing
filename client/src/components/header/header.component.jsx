import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { ReactComponent as Logo } from '../../assets/logo.svg';
import CartIcon from '../cart-icon/cart-icon.component';
import CartDropdownContainer from '../cart-dropdown/cart-dropdown.container';
import { selectCurrentUser } from '../../redux/user/user.selector';
import { selectHidden } from '../../redux/cart/cart.selector';
import { signOutStart } from '../../redux/user/user.action';
import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionDiv,
  OptionLink,
} from './header.style';

export const Header = ({ currentUser, hidden, signOutStart }) => (
  <HeaderContainer>
    <LogoContainer to="/">
      <Logo className="logo" />
    </LogoContainer>
    <OptionsContainer>
      <OptionLink to="shop">Shop</OptionLink>
      <OptionDiv event="none">Contact</OptionDiv>
      {currentUser ? (
        <OptionDiv className="sign-out" onClick={signOutStart}>
          Sign Out
        </OptionDiv>
      ) : (
        <OptionLink className="sign-in" to="signin">
          Sign In
        </OptionLink>
      )}
      <CartIcon />
    </OptionsContainer>
    {hidden ? null : <CartDropdownContainer />}
  </HeaderContainer>
);

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectHidden,
});

const mapDispatchToProps = (dispatch) => ({
  signOutStart: () => dispatch(signOutStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
