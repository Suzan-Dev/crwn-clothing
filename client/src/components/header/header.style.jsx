import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';

const OptionContainerStyles = css`
  padding: 5px;
  cursor: pointer;

  @media screen and (min-width: 800px) {
    padding: 10px 15px;
  }
`;

export const HeaderContainer = styled.div`
  height: 70px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  margin-bottom: 5px;

  @media screen and (min-width: 800px) {
    margin-bottom: 25px;
  }
`;

export const LogoContainer = styled(Link)`
  height: 100%;
  width: 40px;
  padding-top: 10px;
  padding-left: 5px;

  @media screen and (min-width: 800px) {
    padding: 15px;
    width: 70px;
  }
`;

export const OptionsContainer = styled.div`
  width: 80%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;

  @media screen and (min-width: 800px) {
    width: 50%;
  }
`;

export const OptionLink = styled(Link)`
  font-size: 16.5px;
  ${OptionContainerStyles}

  @media screen and (min-width: 800px) {
    font-size: 18px;
  }
`;

export const OptionDiv = styled.div`
  font-size: 16.5px;
  ${OptionContainerStyles}

  @media screen and (min-width: 800px) {
    font-size: 18px;
  }
`;
