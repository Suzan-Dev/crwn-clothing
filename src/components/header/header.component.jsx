import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import './header.style.scss';

import { ReactComponent as Logo } from '../../assets/logo.svg';
import { auth } from '../../firebase/firebase.utils';
import CartIcon from '../cart-icon/cart-icon.component';
import CartDropdown from '../cart-dropdown/cart-dropdown.component';

import { selectCurrentUser } from '../../redux/user/user.selector';
import { selectHidden } from '../../redux/cart/cart.selector';

const Header = ({ currentUser, hidden }) => (
  <div className='header'>
    <Link className='logo-container' to='/'>
      <Logo className='logo' />
    </Link>
    <div className='options'>
      <Link className='option' to='/shop'>
        Shop
      </Link>
      <Link className='option' to='/contact'>
        Contact
      </Link>
      {currentUser ? (
        <div className='option link' onClick={() => auth.signOut()}>
          Sign Out
        </div>
      ) : (
        <Link className='option' to='/signin'>
          Sign In
        </Link>
      )}
      <CartIcon className='option' />
    </div>
    {hidden ? null : <CartDropdown />}
  </div>
);

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectHidden,
});

export default connect(mapStateToProps)(Header);
