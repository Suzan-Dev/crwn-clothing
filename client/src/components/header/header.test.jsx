import React from 'react';
import { shallow } from '../../setupTests';
import { Header } from './header.component';

describe('Testing: Header component!', () => {
  const mockSignOutStart = jest.fn();
  const mockCurrentUser = {
    uid: '1',
  };
  const mockHidden = true;

  const mockProps = {
    signOutStart: mockSignOutStart,
    currentUser: mockCurrentUser,
    hidden: mockHidden,
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Header {...mockProps} />);
  });

  test('should render Header component.', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should not render CartDropdownContainer component.', () => {
    expect(wrapper.exists('CartDropdownContainer')).toBeFalsy();
  });

  describe('If currentUser is present!', () => {
    test('should render Sign Out text.', () => {
      expect(wrapper.find('.sign-out').text()).toBe('Sign Out');
    });

    test('should call signOutStart method when Sign Out is clicked', () => {
      wrapper.find('.sign-out').simulate('click');
      expect(mockSignOutStart).toHaveBeenCalled();
    });
  });

  describe('If currentUser is null!', () => {
    const mockProps2 = {
      signOutStart: mockSignOutStart,
      currentUser: null,
      hidden: mockHidden,
    };
    const wrapper2 = shallow(<Header {...mockProps2} />);

    test('should render Sign In text.', () => {
      expect(wrapper2.find('.sign-in').text()).toBe('Sign In');
    });
  });
});
