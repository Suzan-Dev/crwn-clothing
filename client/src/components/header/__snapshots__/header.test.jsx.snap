// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: Header component! should render Header component. 1`] = `
"<styled.div>
  <Styled(Link) to=\\"/\\">
    <ForwardRef(SvgLogo) className=\\"logo\\" />
  </Styled(Link)>
  <styled.div>
    <Styled(Link) to=\\"shop\\">
      Shop
    </Styled(Link)>
    <styled.div event=\\"none\\">
      Contact
    </styled.div>
    <styled.div className=\\"sign-out\\" onClick={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
      Sign Out
    </styled.div>
    <Connect(CartIcon) />
  </styled.div>
</styled.div>"
`;
